File: packages/ts/src/a2a/showcase.ts
import { compactText } from './text'

/**
 * Versioned identifier for the AI37 Agent Card showcase extension.
 * It is a namespace, not an endpoint: consumers must not dereference it.
 */
export const AI37_SHOWCASE_EXTENSION_URI =
  'https://schemas.ai37.ru/a2a/extensions/showcase/v1' as const

/**
 * A norm the agent computes by. `title` is the full name and is not printed everywhere:
 * compact cards show the code alone. An empty list means the surface writes «Норматив
 * уточняется» — an invented reference is worse than a missing one.
 */
export interface AgentShowcaseNorm {
  code: string
  title?: string
}

/** User-facing description of the agent for the product catalog (page and empty chat screen). */
export interface AgentShowcaseProfile extends Record<string, unknown> {
  title: string
  summary: string
  computes?: string
  norms?: AgentShowcaseNorm[]
  starter?: string
  examples?: string[]
  order?: number
}

export interface AgentShowcaseExtension {
  uri: typeof AI37_SHOWCASE_EXTENSION_URI
  description: string
  required: false
  params: AgentShowcaseProfile
}

const limits = {
  title: 60,
  summary: 160,
  computes: 240,
  starter: 160,
  norms: { items: 4, code: 80, title: 200 },
  examples: { items: 4, length: 160 },
} as const

/**
 * Unlike routing, showcase text is clamped instead of rejected: dropping a whole agent from the
 * catalog over a 61st character would cost the user more than an ellipsis does. The ellipsis is
 * deliberate — a truncated line must look truncated, not like the agent's real name.
 */
function clampText(value: unknown, limit: number): string {
  if (typeof value !== 'string') return ''
  const text = compactText(value)
  if (text.length <= limit) return text
  return `${text.slice(0, limit - 1).trimEnd()}…`
}

function normalizeNorms(value: unknown): AgentShowcaseNorm[] {
  if (!Array.isArray(value)) return []
  const result: AgentShowcaseNorm[] = []
  const seen = new Set<string>()
  for (const item of value) {
    if (result.length >= limits.norms.items) break
    if (!item || typeof item !== 'object' || Array.isArray(item)) continue
    const raw = item as Record<string, unknown>
    const code = clampText(raw.code, limits.norms.code)
    const key = code.toLocaleLowerCase('ru')
    if (!code || seen.has(key)) continue
    seen.add(key)
    const title = clampText(raw.title, limits.norms.title)
    result.push(title ? { code, title } : { code })
  }
  return result
}

function normalizeExamples(value: unknown): string[] {
  if (!Array.isArray(value)) return []
  const result: string[] = []
  const seen = new Set<string>()
  for (const item of value) {
    if (result.length >= limits.examples.items) break
    const example = clampText(item, limits.examples.length)
    const key = example.toLocaleLowerCase('ru')
    if (!example || seen.has(key)) continue
    seen.add(key)
    result.push(example)
  }
  return result
}

/** Optional fields are omitted rather than emitted empty: the card stays readable as JSON. */
function normalizeOptionalFields(
  profile: Record<string, unknown>,
): Omit<AgentShowcaseProfile, 'title' | 'summary'> {
  const optional: Omit<AgentShowcaseProfile, 'title' | 'summary'> = {}
  const computes = clampText(profile.computes, limits.computes)
  if (computes) optional.computes = computes
  const norms = normalizeNorms(profile.norms)
  if (norms.length) optional.norms = norms
  const starter = clampText(profile.starter, limits.starter)
  if (starter) optional.starter = starter
  const examples = normalizeExamples(profile.examples)
  if (examples.length) optional.examples = examples
  if (typeof profile.order === 'number' && Number.isInteger(profile.order)) {
    optional.order = profile.order
  }
  return optional
}

/**
 * Throws only when there is nothing to show: a card that declares the extension without a title or
 * a summary has no place in the catalog, and the agent should learn that on its own CI. Everything
 * else is clamped or dropped.
 */
export function normalizeAgentShowcaseProfile(value: unknown): AgentShowcaseProfile {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new TypeError('showcase profile must be an object')
  }
  const profile = value as Record<string, unknown>
  const title = clampText(profile.title, limits.title)
  const summary = clampText(profile.summary, limits.summary)
  if (!title || !summary) {
    throw new TypeError('showcase.title and showcase.summary are required')
  }
  return { title, summary, ...normalizeOptionalFields(profile) }
}

export function buildAgentShowcaseExtension(
  profile: AgentShowcaseProfile,
): AgentShowcaseExtension {
  return {
    uri: AI37_SHOWCASE_EXTENSION_URI,
    description: 'User-facing showcase profile for the AI37 agent catalog.',
    required: false,
    params: normalizeAgentShowcaseProfile(profile),
  }
}

export function parseAgentShowcaseExtension(
  extensions: readonly unknown[] | null | undefined,
): AgentShowcaseProfile | undefined {
  const extension = extensions?.find(
    (item) =>
      !!item &&
      typeof item === 'object' &&
      (item as Record<string, unknown>).uri === AI37_SHOWCASE_EXTENSION_URI,
  ) as Record<string, unknown> | undefined
  if (!extension) return undefined
  try {
    return normalizeAgentShowcaseProfile(extension.params)
  } catch {
    return undefined
  }
}

File: packages/python/src/ai37_agent_sdk/a2a/showcase.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from typing import Any, NotRequired, TypedDict, cast

from .text import compact_text

# Versioned identifier, not an endpoint. Consumers must not dereference it.
AI37_SHOWCASE_EXTENSION_URI = "https://schemas.ai37.ru/a2a/extensions/showcase/v1"


class AgentShowcaseNorm(TypedDict):
    """Норматив, по которому считает агент. `title` — полное наименование, печатается не везде:
    на компактных карточках остаётся только код. Пустой список означает «Норматив уточняется» —
    выдуманная ссылка хуже отсутствующей."""

    code: str
    title: NotRequired[str]


class AgentShowcaseProfile(TypedDict):
    """Описание агента для витрины продукта: страница «Агенты» и пустой экран чата."""

    title: str
    summary: str
    computes: NotRequired[str]
    norms: NotRequired[list[AgentShowcaseNorm]]
    starter: NotRequired[str]
    examples: NotRequired[list[str]]
    order: NotRequired[int]


class AgentShowcaseExtension(TypedDict):
    uri: str
    description: str
    required: bool
    params: AgentShowcaseProfile


_TITLE_MAX = 60
_SUMMARY_MAX = 160
_COMPUTES_MAX = 240
_STARTER_MAX = 160
_NORMS_MAX_ITEMS = 4
_NORM_CODE_MAX = 80
_NORM_TITLE_MAX = 200
_EXAMPLES_MAX_ITEMS = 4
_EXAMPLE_MAX = 160


def _clamp(value: object, max_length: int) -> str:
    """В отличие от routing текст витрины обрезается, а не отвергается: выкинуть агента из
    каталога из-за 61-го символа дороже для пользователя, чем многоточие. Многоточие намеренное —
    обрезанная строка должна выглядеть обрезанной, а не настоящим названием."""
    if not isinstance(value, str):
        return ""
    text = compact_text(value)
    if len(text) <= max_length:
        return text
    return text[: max_length - 1].rstrip() + "…"


def _norms(value: object) -> list[AgentShowcaseNorm]:
    if not isinstance(value, list):
        return []
    result: list[AgentShowcaseNorm] = []
    seen: set[str] = set()
    for item in value:
        if len(result) >= _NORMS_MAX_ITEMS:
            break
        if not isinstance(item, Mapping):
            continue
        code = _clamp(item.get("code"), _NORM_CODE_MAX)
        if not code or code.casefold() in seen:
            continue
        seen.add(code.casefold())
        title = _clamp(item.get("title"), _NORM_TITLE_MAX)
        result.append({"code": code, "title": title} if title else {"code": code})
    return result


def _examples(value: object) -> list[str]:
    if not isinstance(value, list):
        return []
    result: list[str] = []
    seen: set[str] = set()
    for item in value:
        if len(result) >= _EXAMPLES_MAX_ITEMS:
            break
        example = _clamp(item, _EXAMPLE_MAX)
        if not example or example.casefold() in seen:
            continue
        seen.add(example.casefold())
        result.append(example)
    return result


def _optional_fields(profile: Mapping[str, Any]) -> dict[str, Any]:
    """Пустые опциональные поля не выводим: карточка остаётся читаемой как JSON."""
    optional: dict[str, Any] = {}
    computes = _clamp(profile.get("computes"), _COMPUTES_MAX)
    if computes:
        optional["computes"] = computes
    norms = _norms(profile.get("norms"))
    if norms:
        optional["norms"] = norms
    starter = _clamp(profile.get("starter"), _STARTER_MAX)
    if starter:
        optional["starter"] = starter
    examples = _examples(profile.get("examples"))
    if examples:
        optional["examples"] = examples
    order = profile.get("order")
    if isinstance(order, int) and not isinstance(order, bool):
        optional["order"] = order
    return optional


def normalize_agent_showcase_profile(value: object) -> AgentShowcaseProfile:
    """Падает только когда показывать нечего: карточка, объявившая расширение без заголовка или
    краткого описания, в каталоге не нужна, и агент должен узнать об этом на своём CI. Остальное
    обрезается или отбрасывается."""
    if not isinstance(value, Mapping):
        raise TypeError("showcase profile must be an object")
    profile = cast(Mapping[str, Any], value)
    title = _clamp(profile.get("title"), _TITLE_MAX)
    summary = _clamp(profile.get("summary"), _SUMMARY_MAX)
    if not title or not summary:
        raise TypeError("showcase.title and showcase.summary are required")
    normalized = {"title": title, "summary": summary, **_optional_fields(profile)}
    return cast(AgentShowcaseProfile, normalized)


def build_agent_showcase_extension(profile: AgentShowcaseProfile) -> AgentShowcaseExtension:
    return {
        "uri": AI37_SHOWCASE_EXTENSION_URI,
        "description": "User-facing showcase profile for the AI37 agent catalog.",
        "required": False,
        "params": normalize_agent_showcase_profile(profile),
    }


def parse_agent_showcase_extension(
    extensions: Iterable[object] | None,
) -> AgentShowcaseProfile | None:
    for item in extensions or ():
        if not isinstance(item, Mapping) or item.get("uri") != AI37_SHOWCASE_EXTENSION_URI:
            continue
        try:
            return normalize_agent_showcase_profile(cast(Mapping[str, Any], item).get("params"))
        except (TypeError, ValueError):
            return None
    return None
